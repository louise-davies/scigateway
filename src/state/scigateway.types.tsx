import { Step } from 'react-joyride';
import { Theme } from '@material-ui/core/styles/createMuiTheme';

export const AuthSuccessType = 'scigateway:auth_success';
export const AuthFailureType = 'scigateway:auth_failure';
export const LoadingAuthType = 'scigateway:loading_auth';
export const LoadedAuthType = 'scigateway:loaded_auth';
export const LoadAuthProviderType = 'scigateway:loading_auth_provider';
export const LoadScheduledMaintenanceStateType =
  'scigateway:load_scheduled_maintenance_state';
export const LoadMaintenanceStateType = 'scigateway:load_maintenance_state';
export const InvalidateTokenType = 'scigateway:api:invalidate_token';
export const ConfigureStringsType = 'scigateway:configure_strings';
export const LoginType = 'scigateway:login';
export const NotificationType = 'scigateway:api:notification';
export const RegisterRouteType = 'scigateway:api:register_route';
export const RequestPluginRerenderType = 'scigateway:api:plugin_rerender';
export const SendThemeOptionsType = 'scigateway:api:send_themeoptions';
export const LoadDarkModePreferenceType =
  'scigateway:load_dark_mode_preference';
export const SignOutType = 'scigateway:signout';
export const ToggleDrawerType = 'scigateway:toggledrawer';
export const DismissNotificationType = 'scigateway:dismissnotification';
export const ConfigureFeatureSwitchesType = 'scigateway:feature_switches';
export const AddNotificationsType = 'scigateway:add_notification';
export const SiteLoadingType = 'scigateway:site_loading';
export const ConfigureAnalyticsType = 'scigateway:configure_analytics';
export const InitialiseAnalyticsType = 'scigateway:initialise_analytics';
export const ToggleHelpType = 'scigateway:toggle_help';
export const AddHelpTourStepsType = 'scigateway:add_help_tour_steps';
export const RegisterStartUrlType = 'scigateway:register_start_url';

export interface NotificationPayload {
  message: string;
  severity: string;
}

export interface AddNotificationsPayload {
  notifications: NotificationPayload[];
}

export interface AppStrings {
  [id: string]: string;
}

export interface ConfigureStringsPayload {
  res: ApplicationStrings;
}

export interface ApplicationStrings {
  [section: string]: AppStrings;
}

export interface FeatureSwitchesPayload {
  switches: FeatureSwitches;
}

export interface FeatureSwitches {
  showContactButton: boolean;
  showHelpPageButton: boolean;
}

export interface StartUrlPayload {
  startUrl: string;
}

export interface RegisterRoutePayload {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface PluginConfig {
  section: string;
  link: string;
  plugin: string;
  displayName: string;
  order: number;
  helpText?: string;
  helpSteps?: { target: string; content: string }[];
}

export interface GroupedPlugins {
  [section: string]: PluginConfig[];
}

export interface LoginPayload {
  username: string;
  password: string;
}

export interface AccessPayload {
  authorisation: boolean;
  accessToken: string;
  error: null;
}

export interface AuthorisedPayload {
  token: string;
}

export interface SendThemeOptionsPayload {
  theme: Theme;
  broadcast: boolean;
}

export interface LoadDarkModePreferencePayload {
  darkMode: boolean;
}

export interface DismissNotificationPayload {
  index: number;
}

export interface AuthProviderPayload {
  authProvider: string;
  authUrl: string | undefined;
}

export interface ScheduledMaintenanceStatePayLoad {
  scheduledMaintenance: ScheduledMaintenanceState;
}

export interface ScheduledMaintenanceState {
  show: boolean;
  message: string;
}

export interface MaintenanceStatePayLoad {
  maintenance: MaintenanceState;
}

export interface MaintenanceState {
  show: boolean;
  message: string;
}

export interface SiteLoadingPayload {
  loading: boolean;
}

export interface ConfigureAnalyticsPayload {
  id: string;
}

export interface AddHelpTourStepsPayload {
  steps: Step[];
}
