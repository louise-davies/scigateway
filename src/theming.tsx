import { createMuiTheme, MuiThemeProvider } from '@material-ui/core/styles';
import { ThemeOptions, Theme } from '@material-ui/core/styles/createMuiTheme';
import 'react-redux-toastr/lib/css/react-redux-toastr.min.css';
import React from 'react';
import { StateType } from './state/state.types';
import { connect, useSelector } from 'react-redux';

export interface UKRIThemeOptions extends ThemeOptions {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

export interface UKRITheme extends Theme {
  ukri: {
    bright: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
    contrast: {
      orange: string;
      red: string;
      grey: string;
      blue: string;
    };
    deep: {
      orange: string;
      yellow: string;
      green: string;
      blue: string;
      purple: string;
      red: string;
    };
  };
  drawerWidth: number;
  link: {
    default: string;
    visited: string;
    active: string;
  };
}

/* Colours that may be used across light/dark modes e.g. the main app bar */
const STATIC_COLOURS = {
  darkBlue: '#003088',
  orange: '#FF6900',
};

/* Main colours used for dark/light modes respectively */
interface ThemeColours {
  primary: string;
  secondary: string;
  background: string;
  paper: string;
  blue: string;
  orange: string;
  red: string;
  grey: string;
}

const DARK_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

//For experimenting
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const DARK_MODE_COLOURS_HIGH_CONTRAST: ThemeColours = {
  primary: '#86B4FF',
  secondary: '#80ACFF',
  background: '#1B1B1B',
  paper: '#3A3A3A',
  blue: '#86B4FF',
  orange: '#C34F00',
  red: '#FF7F73',
  grey: '#A4A4A4',
};

const LIGHT_MODE_COLOURS: ThemeColours = {
  primary: '#003088',
  secondary: '#003088',
  background: '#FAFAFA',
  paper: '#FFF',
  blue: '#003088',
  orange: '#C34F00',
  red: '#AC1600',
  grey: '#727272',
};

export const buildTheme = (darkModePreference: boolean): Theme => {
  let options: UKRIThemeOptions;
  const colours = darkModePreference ? DARK_MODE_COLOURS : LIGHT_MODE_COLOURS;

  const overrides = {
    MuiLink: {
      root: {
        color: colours.blue,
      },
    },
    MuiTabs: {
      indicator: {
        color: STATIC_COLOURS.darkBlue,
        textDecoration: 'underline',
      },
    },
    MuiFormLabel: {
      root: {
        '&$error': {
          color: colours.red,
        },
        '&$focused': {
          color: colours.blue,
        },
      },
      asterisk: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiBadge: {
      colorPrimary: {
        backgroundColor: STATIC_COLOURS.orange,
      },
    },
    MuiInput: {
      underline: {
        '&$error:after': {
          borderBottomColor: colours.red,
        },
        '&:after': {
          borderBottomColor: colours.blue,
        },
      },
    },
    MuiOutlinedInput: {
      root: {
        '&$error $notchedOutline': {
          borderColor: colours.red,
        },
      },
    },
    MuiFormHelperText: {
      root: {
        '&$error': {
          color: colours.red,
        },
      },
    },
    MuiPickersToolbar: {
      toolbar: {
        backgroundColor: STATIC_COLOURS.darkBlue,
      },
    },
    MuiPickersCalendarHeader: {
      dayLabel: {
        color: colours.grey,
      },
    },
    MuiPickersDay: {
      current: {
        color: colours.blue,
      },
      dayDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersYear: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      yearSelected: {
        color: colours.blue,
      },
      yearDisabled: {
        color: colours.grey,
      },
    },
    MuiPickersMonth: {
      root: {
        '&:active': {
          color: colours.blue,
        },
        '&:focus': {
          color: colours.blue,
        },
      },
      monthSelected: {
        color: colours.blue,
      },
      monthDisabled: {
        color: colours.grey,
      },
    },
  };

  if (darkModePreference) {
    options = {
      palette: {
        // Light/dark mode
        type: 'dark',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#257fff',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  } else {
    options = {
      palette: {
        // Light/dark mode
        type: 'light',
        primary: {
          main: colours.primary,
        },
        secondary: {
          main: colours.secondary,
        },
        background: {
          default: colours.background,
          paper: colours.paper,
        },
      },
      ukri: {
        bright: {
          orange: '#FF6900', // pure orange
          yellow: '#FBBB10', // yellow
          green: '#67C04D', // light green
          blue: '#1E5DF8', // blue
          purple: '#BE2BBB', // bright purple
          red: '#E94D36', // light red
        },
        contrast: {
          orange: colours.orange,
          red: colours.red,
          grey: colours.grey,
          blue: colours.blue,
        },
        deep: {
          orange: '#C13D33', // pure orange
          yellow: '#F08900', // vivid yellow
          green: '#3E863E', // green
          blue: '#003088', // blue
          purple: '#8A1A9B', // bright purple
          red: '#A91B2E', // red
        },
      },
      drawerWidth: 300,
      link: {
        default: '#1E5DF8',
        visited: '#BE2BBB',
        active: '#E94D36',
      },
      overrides: overrides,
    };
  }

  return createMuiTheme(options);
};

function mapThemeProviderStateToProps(
  state: StateType
): { prefersDarkMode: boolean } {
  return {
    prefersDarkMode: state.scigateway.darkMode,
  };
}

const SciGatewayThemeProvider = (props: {
  children: React.ReactNode;
}): React.ReactElement<{
  children: React.ReactNode;
}> => {
  const darkModePreference: boolean = useSelector(
    (state: StateType) => state.scigateway.darkMode
  );
  return (
    <MuiThemeProvider theme={buildTheme(darkModePreference)}>
      {props.children}
    </MuiThemeProvider>
  );
};

export const ConnectedThemeProvider = connect(mapThemeProviderStateToProps)(
  SciGatewayThemeProvider
);
