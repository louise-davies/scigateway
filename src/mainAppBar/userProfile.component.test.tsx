import React from 'react';
import UserProfileComponent from './userProfile.component';
import { StateType } from '../state/state.types';
import configureStore from 'redux-mock-store';
import { authState, initialState } from '../state/reducers/scigateway.reducer';
import { Provider } from 'react-redux';
import { push } from 'connected-react-router';
import { ThemeProvider, StyledEngineProvider } from '@mui/material';
import thunk from 'redux-thunk';
import TestAuthProvider from '../authentication/testAuthProvider';
import { buildTheme } from '../theming';
import { mount, shallow, ShallowWrapper } from 'enzyme';

describe('User profile component', () => {
  let mockStore;
  let state: StateType;
  const theme = buildTheme(false);

  beforeEach(() => {
    mockStore = configureStore([thunk]);
    state = {
      scigateway: { ...initialState, authorisation: { ...authState } },
    };
    state.scigateway.authorisation.provider = new TestAuthProvider(
      'test-token'
    );
  });

  const createShallowWrapper = (): ShallowWrapper => {
    return shallow(<UserProfileComponent store={mockStore(state)} />)
      .dive()
      .dive();
  };

  it('renders sign in button if not signed in', () => {
    state.scigateway.authorisation.provider = new TestAuthProvider(null);

    const wrapper = createShallowWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('redirects to login when sign in is pressed', () => {
    state.scigateway.authorisation.provider = new TestAuthProvider(null);

    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <UserProfileComponent />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    wrapper.find('button').first().simulate('click');

    expect(testStore.getActions().length).toEqual(1);
    expect(testStore.getActions()[0]).toEqual(push('/login'));
  });

  it('renders sign in button if user signed in via autoLogin', () => {
    state.scigateway.authorisation.provider.autoLogin = () => Promise.resolve();

    window.localStorage.__proto__.getItem = jest
      .fn()
      .mockImplementation((name) => (name === 'autoLogin' ? 'true' : null));

    const wrapper = createShallowWrapper();

    expect(wrapper).toMatchSnapshot();
    expect(localStorage.getItem).toBeCalledWith('autoLogin');
  });

  it('renders default avatar if signed in', () => {
    const wrapper = createShallowWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders user avatar if signed in with avatar url', () => {
    state.scigateway.authorisation.provider.user = {
      username: 'test',
      avatarUrl: 'test_url',
    };
    const wrapper = createShallowWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('opens menu when button clicked', () => {
    const wrapper = mount(
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={theme}>
          <UserProfileComponent store={mockStore(state)} />
        </ThemeProvider>
      </StyledEngineProvider>
    );

    expect(wrapper.find('#simple-menu').first().prop('open')).toBeFalsy();

    wrapper.find('button').simulate('click');

    expect(wrapper.find('#simple-menu').first().prop('open')).toBeTruthy();
  });

  it('signs out if sign out clicked', () => {
    const testStore = mockStore(state);
    const wrapper = mount(
      <Provider store={testStore}>
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={theme}>
            <UserProfileComponent />
          </ThemeProvider>
        </StyledEngineProvider>
      </Provider>
    );

    // Click the user menu button and click on the sign out menu item.
    wrapper.find('button').simulate('click');
    wrapper.find('#item-sign-out').last().simulate('click');

    expect(testStore.getActions().length).toEqual(2);
    expect(testStore.getActions()[0]).toEqual({ type: 'scigateway:signout' });
    expect(testStore.getActions()[1]).toEqual(push('/'));
  });
});
