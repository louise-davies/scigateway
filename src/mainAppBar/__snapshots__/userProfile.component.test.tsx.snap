// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User profile component renders default avatar if signed in 1`] = `
<div
  className="tour-user-profile"
>
  <div>
    <ForwardRef(IconButton)
      aria-haspopup="true"
      aria-label="Open user menu"
      onClick={[Function]}
      size="large"
      sx={
        Object {
          "color": "primary.contrastText",
          "margin": 1,
        }
      }
    >
      <Memo(ForwardRef(AccountCircleIcon)) />
    </ForwardRef(IconButton)>
    <ForwardRef(Menu)
      anchorEl={null}
      id="simple-menu"
      onClose={[Function]}
      open={false}
    >
      <div
        style={
          Object {
            "paddingBottom": "8px",
            "paddingLeft": "15px",
            "paddingRight": "15px",
            "paddingTop": "8px",
          }
        }
      >
        <ForwardRef(Typography)>
          Signed in as:
        </ForwardRef(Typography)>
        <ForwardRef(Typography)
          sx={
            Object {
              "fontSize": "17px",
              "fontWeight": "bold",
              "paddingTop": "3px",
            }
          }
        >
          anonymous
        </ForwardRef(Typography)>
      </div>
      <ForwardRef(Divider) />
      <ForwardRef(MenuItem)
        id="item-sign-out"
        onClick={[Function]}
      >
        <ForwardRef(ListItemIcon)>
          <Memo(ForwardRef(ExitToAppIcon)) />
        </ForwardRef(ListItemIcon)>
        <ForwardRef(ListItemText)
          primary="logout-button"
        />
      </ForwardRef(MenuItem)>
    </ForwardRef(Menu)>
  </div>
</div>
`;

exports[`User profile component renders sign in button if not signed in 1`] = `
<div
  className="tour-user-profile"
>
  <ForwardRef(Button)
    color="primary"
    onClick={[Function]}
    sx={[Function]}
    variant="contained"
  >
    <ForwardRef(Typography)
      color="inherit"
      noWrap={true}
      sx={
        Object {
          "fontWeight": 500,
          "marginTop": "3px",
        }
      }
    >
      login-button
    </ForwardRef(Typography)>
  </ForwardRef(Button)>
</div>
`;

exports[`User profile component renders sign in button if user signed in via autoLogin 1`] = `
<div
  className="tour-user-profile"
>
  <ForwardRef(Button)
    color="primary"
    onClick={[Function]}
    sx={[Function]}
    variant="contained"
  >
    <ForwardRef(Typography)
      color="inherit"
      noWrap={true}
      sx={
        Object {
          "fontWeight": 500,
          "marginTop": "3px",
        }
      }
    >
      login-button
    </ForwardRef(Typography)>
  </ForwardRef(Button)>
</div>
`;

exports[`User profile component renders user avatar if signed in with avatar url 1`] = `
<div
  className="tour-user-profile"
>
  <div>
    <ForwardRef(IconButton)
      aria-haspopup="true"
      aria-label="Open user menu"
      onClick={[Function]}
      size="small"
      sx={
        Object {
          "color": "primary.contrastText",
          "margin": 1,
        }
      }
    >
      <ForwardRef(Avatar)
        alt="user"
        src="test_url"
      />
    </ForwardRef(IconButton)>
    <ForwardRef(Menu)
      anchorEl={null}
      id="simple-menu"
      onClose={[Function]}
      open={false}
    >
      <div
        style={
          Object {
            "paddingBottom": "8px",
            "paddingLeft": "15px",
            "paddingRight": "15px",
            "paddingTop": "8px",
          }
        }
      >
        <ForwardRef(Typography)>
          Signed in as:
        </ForwardRef(Typography)>
        <ForwardRef(Typography)
          sx={
            Object {
              "fontSize": "17px",
              "fontWeight": "bold",
              "paddingTop": "3px",
            }
          }
        >
          test
        </ForwardRef(Typography)>
      </div>
      <ForwardRef(Divider) />
      <ForwardRef(MenuItem)
        id="item-sign-out"
        onClick={[Function]}
      >
        <ForwardRef(ListItemIcon)>
          <Memo(ForwardRef(ExitToAppIcon)) />
        </ForwardRef(ListItemIcon)>
        <ForwardRef(ListItemText)
          primary="logout-button"
        />
      </ForwardRef(MenuItem)>
    </ForwardRef(Menu)>
  </div>
</div>
`;
